### PR TITLE
Work Order Token Processing

### PR DESCRIPTION
**Pull request by mephenor**
```
Added work order token requirement for both endpoints
Added config to provide wps_signing_pubkey 
```

### CODE REVIEW

File: dcs/core/data_repository.py [hunk 1]
```diff
@@ -54,6 +57,10 @@ class DataRepositoryConfig(BaseSettings):
         description="Expiration time in seconds for presigned URLS. Positive integer required",
         example=30,
     )
+    wps_signing_pubkey: str = Field(
+        ...,
+        description="Base64 encoded public key of the keypair that signed the work order tokens",
```
Comment (Cito) [inline dcs/core/data_repository.py:62]: Not base64 encoded, but JSON string.

**Review by Cito — CHANGES_REQUESTED** [review]
General remark: If you use the auth.jwt_auth module from the service commons lib you don't need to deal with the jwcrypto lib directly. It's better to have the crypt and JWT stuff centralized and wrapped in the commons library. But you can also do that as a separate PR or put it on the refactoring list if it takes too much time to refactor now.

One thing that needs to be changed is how the signing key in the config are interpreted, because they are passed as JSON, not as base64 format, see comments.

Otherwise looks good to me, well done.

**Comment by mephenor** [comment]
Now using the Security + AuthContext for token validation.
Had to subclass AuthConfig as setting claims_to_check through the config option behaved weirdly, i.e. still expected name and exp suddenly had an issue with not being an int (?).
Simplified tests as I would need to hook into Security to map validation errors to those I had before, so this is now one error.

Happy path seems to have an issue with the DRS endpoint not being available all the time.
Seems like a timing issue as it seems to happen more frequently in actions than locally.
Might need to introduce a readiness probe there, but that is something for the refactoring log right now

File: tests/test_edge_cases.py [hunk 2]
```diff
@@ -39,6 +42,41 @@ async def test_access_non_existing(joint_fixture: JointFixture):  # noqa F811
...
+    # test with authorization header but wrong pubkey
+    response = await joint_fixture.rest_client.get(
+        f"/objects/{file_id}",
+        timeout=5,
+        headers={"Authorization": f"Bearer {work_order_token}"},
+    )
+    assert response.status_code == status.HTTP_403_FORBIDDEN
+
+    # update pubkey config option on live object
+    auth_provider_override = auth_provider(config=WorkOrderTokenConfig(auth_key=pubkey))
```
[the remark this answers is not in the document]
Comment (Cito) [reply tests/test_edge_cases.py:66]: Then you can also create more specific tests for the edge cases, one for the various cases of 403 when something is wrong with the WOT, and one for the 404 error if the WOT validates.

File: tests/test_typical_journey.py [hunk 3]
```diff
@@ -40,6 +42,20 @@ async def test_happy(
...
+    # simplify testing by using one longer lived work order token
+    work_order_token, pubkey = get_work_order_token(  # noqa: F405
+        file_id=drs_id,
+        user_pubkey=user_pubkey,
+        valid_seconds=120,
+    )
+
+    # modify default headers and patch signing pubkey
+    # mute mypy false positive, dict[str, str] should be compatible with Mapping[str, str]
+    joint_fixture.rest_client.headers = {"Authorization": f"Bearer {work_order_token}"}  # type: ignore
```
Comment (Cito) [inline tests/test_typical_journey.py:56]:
```suggestion
    joint_fixture.rest_client.headers = Headers(
        {"Authorization": f"Bearer {work_order_token}"})
```

And `from httpx import Headers` at the top.

File: tests/test_typical_journey.py [hunk 4]
```diff
@@ -40,6 +42,20 @@ async def test_happy(
...
+
+    # simplify testing by using one longer lived work order token
+    work_order_token, pubkey = get_work_order_token(  # noqa: F405
+        file_id=drs_id,
+        user_pubkey=user_pubkey,
+        valid_seconds=120,
+    )
+
+    # modify default headers and patch signing pubkey
+    # mute mypy false positive, dict[str, str] should be compatible with Mapping[str, str]
```
Comment (Cito) [inline tests/test_typical_journey.py:55]: It expects a special subclass of dict, see below.

File: dcs/adapters/inbound/fastapi_/routes.py [hunk 5]
```diff
@@ -116,29 +139,40 @@ async def get_drs_object(
...
-    "/objects/{object_id}/envelopes/{public_key}",
+    "/objects/{object_id}/envelopes",
     summary="Returns base64 encoded, personalized file envelope",
     operation_id="getEnvelope",
     tags=["DownloadControllerService"],
     status_code=status.HTTP_200_OK,
     response_model=http_response_models.EnvelopeResponseModel,
     response_description="Successfully delivered envelope.",
     responses={
+        status.HTTP_403_FORBIDDEN: RESPONSES["tokenAuthenticationError"],
```
[the remark this answers is not in the document]
Comment (mephenor) [reply dcs/adapters/inbound/fastapi_/routes.py]: I'd leave this as is for now with the custom file error and move this to refactoring until we have talked about httpyexpect

File: dcs/adapters/inbound/fastapi_/http_authorization.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+
+@inject
+async def require_work_order_context(
+    credentials: HTTPAuthorizationCredentials = Depends(HTTPBearer(auto_error=True)),
+    auth_provider: AuthContextProtocol[WorkOrderContext] = Depends(
+        Provide[Container.auth_provider]
+    ),
+) -> WorkOrderContext:
+    """Require an access token using FastAPI."""
```
Comment (Cito) [inline dcs/adapters/inbound/fastapi_/http_authorization.py:37]:
```suggestion
    """Require a work order context using FastAPI."""
```

File: dcs/adapters/inbound/fastapi_/http_authorization.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    return await require_auth_context_using_credentials(
+        credentials=credentials, auth_provider=auth_provider
+    )
+
+
+require_work_order_token = Security(require_work_order_context)
```
Comment (Cito) [inline dcs/adapters/inbound/fastapi_/http_authorization.py:43]:
```suggestion
security_require_work_order_context = Security(require_work_order_context)
```

Or some different name, but it should end with `_context` since it also returns context object, not a token as a string.

File: dcs/core/auth_policies.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        ...,
+        title="Full user name",
+        help="The full name of the user (with academic title)",
+    )
+    email: EmailStr = Field(..., title="E-Mail", help="The email address of the user")
+
+    @validator("type")
+    @classmethod
+    def type_must_be_download(cls, work_type):
+        """Make sure download type matches expectation"""
```
Comment (Cito) [inline dcs/core/auth_policies.py:48]:
```suggestion
        """Make sure work type matches expectation for the download controller"""
```

File: dcs/core/auth_policies.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    )
+    full_user_name: str = Field(
+        ...,
+        title="Full user name",
+        help="The full name of the user (with academic title)",
+    )
+    email: EmailStr = Field(..., title="E-Mail", help="The email address of the user")
+
+    @validator("type")
+    @classmethod
```
Comment (Cito) [inline dcs/core/auth_policies.py:46]: Can you double check that this works? In the FastAPI documentation the classmethod decorator is not used because it's already contained in the validator decorator.

If you're sure that it works, then leave it like that. Otherwise you can silence the linter with
 `# pylint: disable=no-self-argument`.

File: .devcontainer/.dev_config.yaml [hunk 10]
```diff
@@ -35,3 +35,4 @@ files_to_register_topic: internal_file_registry
 files_to_register_type: file_registered
 unstaged_download_event_topic: file_downloads
 unstaged_download_event_type: unstaged_download_requested
+auth_key: "placeholder"
```
Comment (Cito) [inline .devcontainer/.dev_config.yaml:38]:
```suggestion
auth_key: "{}"
```

Just to be consistent with the existing services and to indicate that it's a JSON object.

File: tests/fixtures/joint.py [hunk 11]
```diff
@@ -55,6 +58,9 @@
     decryption_secret_id="some-secret",
 )
 
+SignedToken = str
+PubKey = str
+
 
 def get_free_port() -> int:
```
Comment (Cito) [inline tests/fixtures/joint.py:65]: By the way, I think this function is not needed, since the AsyncClient uses ASGI to access the application and does not use the network. And instead of httpx.AsyncClient you can also use the AsyncTestClient that Kersten has put into the commons lib in v0.3.0 (it's essentially the same, just presets the base url).

File: tests/fixtures/joint.py [hunk 12]
```diff
@@ -63,6 +69,34 @@ def get_free_port() -> int:
...
+
+    jwk = jwt_helpers.generate_jwk()
+
+    signed_token = jwt_helpers.sign_and_serialize_token(
+        claims=claims, key=jwk, valid_seconds=valid_seconds
+    )
+    signing_pubkey = jwk.export_public()
+    return signed_token, signing_pubkey
+
+
```
Comment (Cito) [inline tests/fixtures/joint.py:99]: Yes. that's better.

**Review by Cito — APPROVED** [review]
Thanks, this should be good enough. Just some more very minor suggestions.

File: dcs/core/auth_policies.py (hunk 9 above)
Comment (mephenor) [reply dcs/core/auth_policies.py:46]: Also works this way, but changed it to not be unnecessary redundant

File: dcs/adapters/inbound/fastapi_/http_authorization.py (hunk 7 above)
Comment (mephenor) [reply dcs/adapters/inbound/fastapi_/http_authorization.py:43]: I like the second solution

File: tests/fixtures/joint.py (hunk 11 above)
Comment (mephenor) [reply tests/fixtures/joint.py:65]: Seems to work, changed accordingly

**Review by Cito — APPROVED** [review]
Your last commit message sounds like I should not request any more changes :)

**Comment by mephenor** [comment]
That would be nice for now :D
Though that was actually referring to the testing issue,  I haven't found the underlying cause yet, so this might creep up again
